rust/src/tracing.rs

```diff
@@ -0,0 +1,362 @@
+//! in your terminal's standard output, register the [`TracingLogListener`] after initializing your
+//! tracing subscriber.
+//!
+//! ## Never use both [`BinaryNinjaLayer`] and [`TracingLogListener`] simultaneously
+//!
+//! Enabling both creates an infinite feedback loop where a log triggers a log, deadlocking the application.
+
+use crate::file_metadata::SessionId;
+use crate::logger::{
+    bn_log_with_session, BnLogLevel, LogContext, LogGuard, LogListener, LOGGER_DEFAULT_SESSION_ID,
+};
+use tracing::{Event, Id, Level, Subscriber};
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::prelude::*;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::Layer;
+
+/// Helper macro to initialize the [`BinaryNinjaLayer`] tracing layer for plugins.
+///
+/// Maps the current crate name to the provided display name and enables target formatting.
+/// Use [`init_with_layer`] if you intend on registering a [`BinaryNinjaLayer`] with non-default values.
+///
+/// ## Note for Plugins
+///
+/// This should **only** be called once, at the start of plugins.
+///
+/// ```no_run
+/// #[unsafe(no_mangle)]
+/// pub unsafe extern "C" fn CorePluginInit() -> bool {
+///     binaryninja::tracing_init!("MyPlugin");
+///     tracing::info!("Core plugin initialized");
+///     true
+/// }
+/// ```
+///
+/// ## Note for Headless
+///
+/// This should **never** be called if you are running headlessly and as you will likely be
+/// registering a [`LogListener`], which will possibly round-trip back to tracing logs and deadlock
+/// the program.
+#[macro_export]
+macro_rules! tracing_init {
+    () => {
+        let layer = $crate::tracing::BinaryNinjaLayer::default();
+        $crate::tracing::init_with_layer(layer);
+    };
+    ($name:expr) => {
+        let layer = $crate::tracing::BinaryNinjaLayer::default()
+            .with_target_mapping(env!("CARGO_CRATE_NAME"), $name);
+        $crate::tracing::init_with_layer(layer);
+    };
+}
+
+/// Initialize the core tracing subscriber with the given [`BinaryNinjaLayer`]. Collects and sends logs
+/// to the core.
+pub fn init_with_layer(layer: BinaryNinjaLayer) {
+    let subscriber = tracing_subscriber::registry().with(layer);
+    let _ = tracing::subscriber::set_global_default(subscriber);
+}
+
+/// Subscribes to all spans and events emitted by `tracing` and forwards them to the core logging API.
+#[derive(Clone)]
+pub struct BinaryNinjaLayer {
+    /// Rewrite mappings for the default target.
+    ///
+    /// # Example
+    ///
+    /// Given the target "my_crate::commands::analyze" and the mapping ("my_crate", "MyPlugin") the
+    /// target will be rewritten to "MyPlugin::commands::analyze", assuming no other rewrites occur.
+    pub target_mappings: Vec<(String, String)>,
+    /// Whether the default target should be formatted to be displayed in the "common" logger name
+    /// format for Binary Ninja. This formatting does _not_ apply to targets explicitly set, rather
```

Comment:
It might be clearer to phrase this in the positive. Something like: "This formatting will only be applied when the target name is implicit. Explicitly provided target names will be preserved as-is."